### PR TITLE
federation service controller: fixing the logic to update DNS records

### LINKED ISSUE #27989: Deleting a backend deletes too many DNS records
Step 1.  Create a two-cluster federation, one in zone A one in F.  Create a loadbalancer service in the federation, and create one pod in each cluster that matches the service selector.  Result is good:

```
NAME                                                                                                 TYPE  TTL    DATA
kube.5yetis.net.                                                                                     NS    21600  ns-cloud-b1.googledomains.com., ns-cloud-b2.googledomains.com., ns-cloud-b3.googledomains.com., ns-cloud-b4.googledomains.com.
kube.5yetis.net.                                                                                     SOA   21600  ns-cloud-b1.googledomains.com. cloud-dns-hostmaster.google.com. 4 21600 3600 259200 300
federated-service.e2e-tests-service-lt9rd.federation.svc.kube.5yetis.net.                            A     180    104.197.202.248, 23.236.61.35
federated-service.e2e-tests-service-lt9rd.federation.svc.us-central1.kube.5yetis.net.                A     180    104.197.202.248, 23.236.61.35
federated-service.e2e-tests-service-lt9rd.federation.svc.us-central1-a.us-central1.kube.5yetis.net.  A     180    104.197.202.248
federated-service.e2e-tests-service-lt9rd.federation.svc.us-central1-f.us-central1.kube.5yetis.net.  A     180    23.236.61.35
```

Step 2.  Delete the pod in F.

```console
% kc --context=federation-e2e-gce-us-central1-f delete pod backend --namespace=e2e-tests-service-lt9rd
pod "backend" deleted
```

Step 3. Zone file looks very sad.  The remaining A record feels very lonely.  :sob:
```
NAME                                                                                                 TYPE  TTL    DATA
kube.5yetis.net.                                                                                     NS    21600  ns-cloud-b1.googledomains.com., ns-cloud-b2.googledomains.com., ns-cloud-b3.googledomains.com., ns-cloud-b4.googledomains.com.
kube.5yetis.net.                                                                                     SOA   21600  ns-cloud-b1.googledomains.com. cloud-dns-hostmaster.google.com. 4 21600 3600 259200 300
federated-service.e2e-tests-service-lt9rd.federation.svc.us-central1-a.us-central1.kube.5yetis.net.  A     180    104.197.202.248
```

[federation-controller-manager.log.txt](https://github.com/kubernetes/kubernetes/files/330980/federation-controller-manager.log.txt)

@quinton-hoole @kubernetes/sig-cluster-federation

### PR DESCRIPTION
**Pull request by ghost**
https://github.com/kubernetes/kubernetes/commit/dd78dd8e2b623d32b57a084a855a2a0327d55f38 introduced an 'endpointsReachable' boolean argument to ensureDnsRrrsets(), which is incorrect.

The logic prior to the above commit does the right thing, provided that the correct set of healthy endpoints are passed into that function.  

This PR essentially reverts the part of the above commit related to 'endpointsReachable', and adds some debug logging (at --v=4) in case it becomes necessary.

cc: @mfanjie @mml @nikhiljindal @madhusudancs

### CODE REVIEW

File: federation/pkg/federation-controller/service/dns.go [hunk 1]
```diff
@@ -164,86 +166,90 @@ func (s *ServiceController) ensureDnsRrsets(dnsZoneName, dnsName string, endpoin
...
-					return err // TODO: We could potentially add the ones we did get back, even if some of them failed to resolve.
-				}
-				newRrset := rrsets.New(dnsName, resolvedEndpoints, minDnsTtl, rrstype.A)
+		glog.V(4).Infof("No recordsets found for DNS name %q.  Need to add either A records (if we have healthy endpoints), or a CNAME record to %q", dnsName, uplevelCname)
+		if len(endpoints) < 1 {
+			glog.V(4).Infof("There are no healthy endpoint addresses at level %q, so CNAME to %q, if provided", dnsName, uplevelCname)
+			if uplevelCname != "" {
+				glog.V(4).Infof("Creating CNAME to %q for %q", uplevelCname, dnsName)
+				newRrset := rrsets.New(dnsName, []string{uplevelCname}, minDnsTtl, rrstype.CNAME)
 				rrset, err = rrsets.Add(newRrset)
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/service/dns.go]: Add `glog.V(4).Infof("Adding recordset %v", newRrset)` as you do in the else block.

File: federation/pkg/federation-controller/service/dns.go [hunk 2]
```diff
@@ -164,86 +166,90 @@ func (s *ServiceController) ensureDnsRrsets(dnsZoneName, dnsName string, endpoin
...
+			glog.V(4).Infof("Adding recordset %v", newRrset)
+			rrset, err = rrsets.Add(newRrset)
+			if err != nil {
+				return err
+			}
+			glog.V(4).Infof("Successfully added recordset %v", newRrset)
 		}
 	} else {
 		// the rrset already exists, so make it right.
+		glog.V(4).Infof("Recordset %v already exists.  Ensuring that it is correct.")
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/service/dns.go]: Missing param for %v.
I have had verify script failed on my PRs for such an error. Am not sure why its not failing on this PR.
I think its verify/go-vet that fails.

File: federation/pkg/federation-controller/service/dns.go [hunk 3]
```diff
@@ -164,86 +166,90 @@ func (s *ServiceController) ensureDnsRrsets(dnsZoneName, dnsName string, endpoin
...
+					glog.V(4).Infof("Successfully added needed recordset %v", newRrset)
+				} else {
+					glog.V(4).Infof("Uplevel CNAME is empty string. Not adding recordset %v", newRrset)
 				}
 			}
 		} else {
 			// We have an rrset in DNS, possibly with some missing addresses and some unwanted addresses.
 			// And we have healthy endpoints.  Just replace what's there with the healthy endpoints, if it's not already correct.
+			glog.V(4).Infof("%s: Healthy endpoints %v exist.  Recordset %v exists.  Reconciling.", dnsName, endpoints, rrset)
 			resolvedEndpoints, err := getResolvedEndpoints(endpoints)
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/service/dns.go]: What if len(resolvedEndpoints) == 0? Is it still fine to add?
(In comparison, in the other if block, we check uplevelCname != "" before adding).

**Comment by nikhiljindal** [comment]
Thanks for adding the logs. They are going to be super useful.
I am surprised you didnt have to update tests.
We should add tests for this code :)


Adding LGTM. You can reply to comments/ resolve them in follow up PR.

**Comment by nikhiljindal** [comment]
Fixed release note and title

File: federation/pkg/federation-controller/service/dns.go (hunk 2 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/service/dns.go]: Interesting.  I ran go vet on this manually, and it passed. Maybe I omitted some arguments that verify/go-vet does.  Either way, I'll fix this now.

File: federation/pkg/federation-controller/service/dns.go (hunk 1 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/service/dns.go]: Agreed, done. in followup PR.

File: federation/pkg/federation-controller/service/dns.go (hunk 2 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/service/dns.go]: Done in followup PR.

File: federation/pkg/federation-controller/service/dns.go (hunk 3 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/service/dns.go]: In this else clause, len(endpoints) is by definition > 0.  And we're adding an A record, so the value of uplevelCname in this case is irrelevant, as we're not adding a CNAME record.  Make sense @nikhiljindal ?

**Pull request merged**